File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 100000
    l, u = 1, 100000
    
    n = len(diffs)
    while l < u:
        m = (l+u) // 2
        time = 0
        
        for i in range(n):
            if diffs[i] <= m:
                time += times[i]
            else:
                time += (diffs[i] - m) * (times[i]+times[i-1])
                time += times[i]
                
            if time > limit:
                l = m+1
                break
        else:
            u = m
            if answer > m:
                answer = m
                
    
    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 100000;
        
        int l = 1;
        int u = 100000;
        int n = times.length;
        
        while (l < u){
            int level = (l+u)/2;
            long t = 0;
            
            boolean check = true;
            for (int i=0; i<n; i++){
                if (diffs[i] <= level){
                    t += times[i];
                } else {
                    t += (diffs[i] - level) * (times[i] + times[i-1]) + times[i];
                }
                
                if (t > limit){
                    check = false;
                    l = level+1;
                    break;
                }
            }
            
            if (check){
                u = level;
                if (answer > level) answer = level;
            }
        }
        
        
        return answer;
    }
}